flow-utilities/index.js

```diff
@@ -0,0 +1,187 @@
+        delay_duration: get(inputs, "delay_duration.value", ""),
+      };
+
+      app.logger.info("flow-utilities: Opening config view", state);
+      await app.client.views.open({
+        token: context.botToken,
+        trigger_id: body.trigger_id,
+        view: renderStepConfig(state),
+      });
+    }
+  );
+
+  // The subtype has been selected
+  app.action(ACTION_SUBTYPE, async ({ body, action, ack, client }) => {
+    ack();
+    const { view } = body;
+    const subtype = action.value
+    const state = {
+      subtype,
+    };
+
+    await client.views.update({
+      view_id: view.id,
+      view: renderStepConfig(state),
+    });    
+
+  })
+
+  // Handle saving of step config
+  app.view(VIEW_CALLBACK_ID, async ({ ack, view, body, client, logger }) => {
+    let metadata = {}
+    try {
+      metadata = JSON.parse(view.private_metadata)
+    } catch(e) {
+      logger.error("flow-utilities: Error parsing private metadata", e)
+    }
+
+    const subtype = metadata.subtype || ''
+    const workflowStepEditId = get(body, `workflow_step.workflow_step_edit_id`);
+    const durationStr = get(view, `state.values.${BLOCK_DURATION}.${ELEMENT_DURATION}.value`);
+
+    // TODO make dyanmic for different subtypes
+    let stepName = 'Delay the workflow';
+
+    // check if duration contains any dynamic references
+    if (durationStr.indexOf("{") < 0) {
+      const durationMs = parse(durationStr) || 0
+      stepName = `Delay the workflow for about ${moment.duration(durationMs).humanize()}`
+    }
+
+    const inputs = {
+      delay_duration: {
+        value: durationStr,
+      },
+      subtype: {
+        value: subtype,
+      }
+    };
+
+    ack();
+
+    const params = {
+      workflow_step_edit_id: workflowStepEditId,
+      inputs,
+      outputs: [],
+      step_name: stepName,
+    };
+
+    logger.info("updateStep params: ", params);
+    try {
+      await client.workflows.updateStep(params);
+    } catch (e) {
+      logger.error("error updating step: ", e.message);
+    }
+  });
+
+  // Handle execution of the step
+  app.event("workflow_step_execute", async ({ event, logger, context }) => {
+    const { callback_id, workflow_step = {} } = event;
+    if (callback_id !== STEP_CALLBACK_ID) {
+      return;
+    }
+
+    const { inputs = {}, workflow_step_execute_id } = workflow_step;
+    const subtype = ( inputs.subtype || {} ).value || ''
+
+    switch (subtype) {
+      case DELAY_SUBTYPE:
+        const durationStr = ( inputs.delay_duration || {} ).value || ''
+        const durationMs = parse(durationStr) || 0
+        const futureUnixTime = moment().add(durationMs, 'ms').unix();
+
+        // TODO: refactor this out when we have a second usecase for the scheduled message queue
+        // The simple `type` and `message` property of the envelop should apply regardless.
+        const scheduled = {
+          type: subtype,
+          message: {
+            step_completed_payload: {
```

Comment:
I'd like to be able to store the team id in this message rather than the token and then lookup the token when processing the message but I can't figure out how to get the team id in this context... yet.

Reply:
You can grab the team id from the full payload, accessible if you add `body` to the args in the handler, then it's `body.team_id`.